### PR TITLE
[DeadCode] Handle crash on RemoveUnusedPrivatePropertyRector+RemoveJustPropertyFetchForAssignRector

### LINKED ISSUE #7495: Crash of RemoveUnusedPrivatePropertyRector, RemoveJustPropertyFetchForAssignRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/b426dfd8-47f1-4ac5-8ddc-0a48b5bcfeed

```php
<?php

class OrderInvoice
{
    private UrlInterface $urlInterface;
    private InvoicesHelper $invoicesHelper;
    private Registry $coreRegistry;

    public function __construct(
        UrlInterface $urlInterface,
        Template\Context $context,
        GetInvoicesHelper $invoicesHelper,
        Registry $registry,
    ) {
        $this->urlInterface   = $urlInterface;
        $this->invoicesHelper = $invoicesHelper;
        $this->coreRegistry   = $registry;

        parent::__construct($context);
    }

    public function getInvoices(): InvoicesHelper
    {
        return $this->invoicesHelper;
    }

    public function getOrder(): Registry
    {
        return $this->coreRegistry;
    }
}
```

that will crash with the following message

```
[20-Sep-2022 22:54:21 Europe/Berlin] PHP Fatal error:  Uncaught TypeError: Rector\DeadCode\NodeAnalyzer\JustPropertyFetchVariableAssignMatcher::matchVariableAndPropertyFetchAssign(): Argument #1 ($stmt) must be of type PhpParser\Node\Stmt, null given, called in /var/www/default/.rector/vendor/rector/rector/rules/DeadCode/NodeAnalyzer/JustPropertyFetchVariableAssignMatcher.php on line 33 and defined in /var/www/default/.rector/vendor/rector/rector/rules/DeadCode/NodeAnalyzer/JustPropertyFetchVariableAssignMatcher.php:51
Stack trace:
#0 /var/www/default/.rector/vendor/rector/rector/rules/DeadCode/NodeAnalyzer/JustPropertyFetchVariableAssignMatcher.php(33): Rector\DeadCode\NodeAnalyzer\JustPropertyFetchVariableAssignMatcher->matchVariableAndPropertyFetchAssign(NULL)
#1 /var/www/default/.rector/vendor/rector/rector/rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchForAssignRector.php(71): Rector\DeadCode\NodeAnalyzer\JustPropertyFetchVariableAssignMatcher->match(Object(PhpParser\Node\Stmt\ClassMethod))
#2 /var/www/default/.rector/vendor/rector/rector/src/Rector/AbstractRector.php(212): Rector\DeadCode\Rector\StmtsAwareInterface\RemoveJustPropertyFetchForAssignRector->refactor(Object(PhpParser\Node\Stmt\ClassMethod))
#3 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode(Object(PhpParser\Node\Stmt\ClassMethod))
#4 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#5 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Class_))
#6 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#7 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Namespace_))
#8 /var/www/default/.rector/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray(Array)
#9 /var/www/default/.rector/vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(42): PhpParser\NodeTraverser->traverse(Array)
#10 /var/www/default/.rector/vendor/rector/rector/src/Application/FileProcessor.php(60): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse(Array)
#11 /var/www/default/.rector/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(133): Rector\Core\Application\FileProcessor->refactor(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#12 /var/www/default/.rector/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(100): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#13 /var/www/default/.rector/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(160): Rector\Core\Application\FileProcessor\PhpFileProcessor->process(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#14 /var/www/default/.rector/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(133): Rector\Core\Application\ApplicationFileProcessor->processFiles(Array, Object(Rector\Core\ValueObject\Configuration))
#15 /var/www/default/.rector/vendor/rector/rector/src/Console/Command/ProcessCommand.php(118): Rector\Core\Application\ApplicationFileProcessor->run(Object(Rector\Core\ValueObject\Configuration), Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput))
#16 /var/www/default/.rector/vendor/rector/rector/vendor/symfony/console/Command/Command.php(307): Rector\Core\Console\Command\ProcessCommand->execute(Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix202209\Symfony\Component\Console\Output\ConsoleOutput))
#17 /var/www/default/.rector/vendor/rector/rector/vendor/symfony/console/Application.php(894): RectorPrefix202209\Symfony\Component\Console\Command\Command->run(Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix202209\Symfony\Component\Console\Output\ConsoleOutput))
#18 /var/www/default/.rector/vendor/rector/rector/vendor/symfony/console/Application.php(310): RectorPrefix202209\Symfony\Component\Console\Application->doRunCommand(Object(Rector\Core\Console\Command\ProcessCommand), Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix202209\Symfony\Component\Console\Output\ConsoleOutput))
#19 /var/www/default/.rector/vendor/rector/rector/src/Console/ConsoleApplication.php(49): RectorPrefix202209\Symfony\Component\Console\Application->doRun(Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix202209\Symfony\Component\Console\Output\ConsoleOutput))
#20 /var/www/default/.rector/vendor/rector/rector/vendor/symfony/console/Application.php(208): Rector\Core\Console\ConsoleApplication->doRun(Object(RectorPrefix202209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix202209\Symfony\Component\Console\Output\ConsoleOutput))
#21 /var/www/default/.rector/vendor/rector/rector/bin/rector.php(128): RectorPrefix202209\Symfony\Component\Console\Application->run()
#22 /var/www/default/.rector/vendor/rector/rector/bin/rector(5): require_once('/var/www/default/.rector/vendor/rector/rector/bin/rector.php')
#23 /var/www/default/.rector/vendor/bin/rector(120): include('/var/www/default/.rector/vendor/rector/rector/bin/rector')
#24 {main}
```

### Responsible rules
* `RemoveUnusedPrivatePropertyRector`
* `RemoveJustPropertyFetchForAssignRector`

## Expected Behavior
Rector should not crash.

I already track down this issue. The problem is, that [`$stmt[0]`](https://github.com/rectorphp/rector-src/blob/36fe5d5d05eae7a4c0ce74513d378d80737c16c8/rules/DeadCode/NodeAnalyzer/JustPropertyFetchVariableAssignMatcher.php#L34) does not exist:
image

I don't know why the array is not ordered correctly. Maybe the rule `RemoveUnusedPrivatePropertyRector` perform a change and a node gets deleted?

I think this can be a common problem for all rules, in case of `...->stmts` usage. So it's possible to fix is globally, or must this fix for all `...->stmts` usage?

If it can't be fixed globally, I would prefer to use a `array_value` to fix the order. Sounds good to you?

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
class Fixture
{
    private UrlInterface $urlInterface;
    private InvoicesHelper $invoicesHelper;
    private Registry $coreRegistry;

    public function __construct(
        UrlInterface $urlInterface,
        Template\Context $context,
        GetInvoicesHelper $invoicesHelper,
        Registry $registry,
    ) {
        $this->urlInterface   = $urlInterface;
        $this->invoicesHelper = $invoicesHelper;
        $this->coreRegistry   = $registry;

        parent::__construct($context);
    }

    public function getInvoices(): InvoicesHelper
    {
        return $this->invoicesHelper;
    }

    public function getOrder(): Registry
    {
        return $this->coreRegistry;
    }
}
```

It crash:

```bash
There was 1 error:

1) Rector\Core\Tests\Issues\Issue7495\RemoveUnusedPrivateJustPropertyFetchTest::test with data set #0 ('/Users/samsonasik/www/rector-...hp.inc')
Undefined array key 0
```

Ref https://getrector.org/demo/b426dfd8-47f1-4ac5-8ddc-0a48b5bcfeed

Fixes https://github.com/rectorphp/rector/issues/7495

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @Wohlie

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: src/Rector/AbstractRector.php [hunk 1]
```diff
@@ -200,6 +200,8 @@ final public function enterNode(Node $node)
 
         $this->printDebugCurrentFileAndRule();
 
+        $this->changedNodeScopeRefresher->reIndexNodeAttributes($node);
+
```
Comment (TomasVotruba) [inline src/Rector/AbstractRector.php:204]: This used to decrease performance quiet a bit.
That's why it should run only when node changed. 

How does it affect speed now?

Comment (samsonasik) [reply src/Rector/AbstractRector.php:204]: This is not refresh Scope, only reindex stmts early, params, closure uses. array_values() should not take many performance degradation, except per node has 1000+ subnodes.

Comment (TomasVotruba) [reply src/Rector/AbstractRector.php:204]: Ok, let's ship it :+1:

**Pull request merged**